### ISSUE TITLE
XML parsing fails for extremely large backups

### ISSUE THREAD

**Issue by theprez1980** [ragibson/SMS-MMS-deduplication#8]
Attempting to run this script on a 40 GB XML file taken from a Galaxy S24 with SMS, RCS and MMS content on Ubuntu.   Running the script generates the following error after several minutes:

Reading 'sms-20241002055532.xml'... Traceback (most recent call last):
  File "sms-20241002055532.xml/dedupe_texts.py", line 296, in <module>
    input_tree = read_input_xml(input_fp)
  File "sms-20241002055532.xml/dedupe_texts.py", line 58, in read_input_xml
    tree = parse(filepath, parser=p)
  File "src/lxml/etree.pyx", line 3536, in lxml.etree.parse
  File "src/lxml/parser.pxi", line 1876, in lxml.etree._parseDocument
  File "src/lxml/parser.pxi", line 1902, in lxml.etree._parseDocumentFromURL
  File "src/lxml/parser.pxi", line 1805, in lxml.etree._parseDocFromFile
  File "src/lxml/parser.pxi", line 1177, in lxml.etree._BaseParser._parseDocFromFile
  File "src/lxml/parser.pxi", line 615, in lxml.etree._ParserContext._handleParseResultDoc
  File "src/lxml/parser.pxi", line 725, in lxml.etree._handleParseResult
  File "src/lxml/parser.pxi", line 654, in lxml.etree._raiseParseError
  File "sms-20241002055532.xml", line 192923
lxml.etree.XMLSyntaxError: AttValue length too long, line 192923, column 1000000233

**Comment by ragibson** [comment]
This crash is on the initial parsing from lxml.

Are you sure this XML isn't corrupted? That error is suggesting you have a line in the XML file that's *one billion* characters long?

**Comment by theprez1980** [comment]
Thanks for the reply and info.   

The XML was exported from an app called "SMS Backup and Restore Pro" on the mobile device, the same app can read and reimport the same file so it appears it's not corrupted as far as I can tell.  

The viewer app found here https://www.synctech.com.au/sms-backup-restore/view-backup/  also can read the file if that helps any.   

Thanks

**Comment by ragibson** [comment]
I think this is probably a bug in lxml. What does that line 192923 look like in `sms-20241002055532.xml`?

**Comment by theprez1980** [comment]
That line shows references to a quicktime video:

**192922** <parts>
**192923**  <part seq="0" ct="video/quicktime" name="EAddw45432223r322a.MOV" ...
**192924** </parts>

**Comment by ragibson** [comment]
Interesting, does the file also crash the standard library parser? What happens if you run
```python
import xml.etree.ElementTree as ET
tree = ET.parse('sms-20241002055532.xml')
print(next(iter(tree.getroot())))
```

**Comment by theprez1980** [comment]
I placed those 3 lines into a test.py file, ran it and it errored out with this:

xml.etree.ElementTree.ParseError: out of memory: line 192923, column 6

The Ubuntu box is running on Proxmox with 100GB of RAM allocated to it, so I'm going to double it and report back.

**Comment by theprez1980** [comment]
I'm guessing there's a memory leak or something else occurring with the parser.  With 300GB allocated to the box, it climbs to about 80GB in memory consumption and then errors out with the ParseError out of memory.   strange.

**Comment by ragibson** [comment]
Does `import xml.etree.cElementTree as ET` work instead? The stack overflow thread seems to indicate it might work better.

**Comment by theprez1980** [comment]
Took longer, but eventually errored out, same line.   Opened up UltraEdit64, removed the associated block, saved changes, tried again.   Errored out with the same out of memory line but a different line.   I'm guessing the file is just too large perhaps.

**Comment by ragibson** [comment]
Yeah, if it crashes all the XML libraries, I'm not sure there's much that can be done.

_Maybe_ iteratively parsing the XML would work, but it would probably require a bit of refactoring.

**Comment by markymark1210a** [comment]
@theprez1980 I encountered the same issue, though my XML file was around 8 GB in size.  One thing you may want to consider, since you're using SMS Backup and Restore, is to break up the backups by date (eg 1/1/2020 to 6/1/2020) and see if you can reduce the size of the XML thay way - you'll have to likely break it apart into many separate XML files to get all the way to 2024.  That worked for me, but your XML could have more photos or videos than mine.  I ended up getting my XMLs to around 1GB using that approach.

@ragibson Iteratively parsing sounds like a great idea.  With RCS now live, I think most people's XML files will simply grow larger than what the current method reliably supports.

Regardless tho - thanks for your time and the script.

**Comment by ragibson** [comment]
It's interesting, I've had people successfully run files up to at least 20GB, so I wonder if the issue is with single elements being unusually large (like sending a video through RCS).

**Comment by janez33** [comment]
@markymark1210a how did you managed to solve the issue?
i have 5gb backup, my backups never work ok, always duplicated, im using Textra as an messaging app
OSError: Error reading file 'sms-20241103222247.x
![image](https://github.com/user-attachments/assets/7caed7cf-3501-4417-a659-be6aa33099c7)
ml': failed to load external entity "sms-20241103222247.xml"

**Comment by janez33** [comment]
@theprez1980 Hello, thank you for all the details :) 
I was on Windows 10.
Now I switched to Linux and voila! It works :) 
![slika](https://github.com/user-attachments/assets/faa95b6f-3b7b-4906-803c-cc6a432c97d7)

**Comment by ragibson** [comment]
Interesting, I wonder why lxml's behavior differs on Windows. The error seemed to indicate that the file didn't exist? Or somehow includes itself as an external reference?

**Comment by ragibson** [comment]
I constructed an XML file that replicates this, and it appears to occur when elements exceed 1 billion characters / 750 MB.

I'm not sure how such elements appear in texting backups since all carrier limits for RCS seem to be closer to 100-200 MB.

Regardless, I've raised an lxml bug report at https://bugs.launchpad.net/lxml/+bug/2101805.

**Comment by theprez1980** [comment]
@ragibson Thanks for the update and for raising this up.  I suspect the limit was exceeded as parts of the message from the backup was from iMessage.

**Comment by ragibson** [comment]
Apparently this is an arbitrary hard security limit set in libxml2, which also means that iteratively parsing the file wouldn't fix it.

That said, I've added instructions on how to build your own versions of libxml2 and lxml that bypass this limit in [handling_extremely_large_text_messages.md](https://github.com/ragibson/SMS-MMS-deduplication/blob/main/handling_extremely_large_text_messages.md), but this is _quite_ a power user thing to do.

I'll close this issue for now as I don't think there's much more that can be done here.

**Comment by ragibson** [comment]
Well, considering the lxml maintainer seemed unaware of this, I've also raised it as a bug in libxml2 at https://gitlab.gnome.org/GNOME/libxml2/-/issues/874, if only for awareness's sake.

**Comment by janez33** [comment]
Hi, sorry for asking here but I didn't know where is the best place to ask. If you need me to move conversation or delete it let me know.
So my problem was before I discovered your repo that my MMS were always duplicated after restoring with SMS Backup & Restore application.  I always thought that's because of me using Textra ap for messaging. So I wiped everything and started fresh on a new phone. Few months ago I discovered your repo finally deduplicated my messages and imported without any problems. So now I'm thinking what if I try to deduplicate some of my oldest backups and import it. Will message history be broken or can I break something? Also what program is ok to open such large backups to figure out dates?  Viewer from SMS Backup & Restore crashes.. hope you can answer my questions, thank you

**Comment by ragibson** [comment]
Yes, it should be okay, but I'd make backups first just in case.

As long as they're consistent, you _should_ be able to restore an old
deduplicated set over your existing messages without issues if you keep the
duplicate check enabled in SMS B&R.
